RAGchain/benchmark/dataset/msmarco.py

```diff
@@ -0,0 +1,115 @@
+        Supporting metrics are 'Recall', 'Precision', 'Hole', 'TopK_Accuracy', 'EM', 'F1_score', 'context_recall', 'context_precision', 'answer_relevancy', 'faithfulness', 'NDCG'.
+        You must ingest all data for using context_recall and context_precision metrics.
+        """
+
+        self.file_path = "ms_marco"
+        # You can available MSMARCO dataset versions v1.1 and v2.1
+        self.dataset = load_dataset(self.file_path, 'v1.1')
+
+        support_metrics = ['Recall', 'Precision', 'Hole', 'TopK_Accuracy', 'EM', 'F1_score', 'context_recall',
+                           'context_precision', 'answer_relevancy', 'faithfulness', 'NDCG']
+        if metrics is not None:
+            using_metrics = list(set(metrics))
+        else:
+            using_metrics = support_metrics
+        super().__init__(run_all=False, metrics=using_metrics)
+
+        self.eval_size = evaluate_size
+        self.run_pipeline = run_pipeline
+        self.retrieval_gt_lst = []
+        self.retrieval_gt_ord_lst = []
+        self.data = self.dataset['test']
+        self.for_passage = pd.DataFrame(
+            {'query_id': self.data['query_id'], 'question': self.data['query'], 'passages': self.data['passages'],
+             'answers': self.data['answers']})
+
+        # Remove non answers and is_selected list elements are all 0 and non answer rows.
+        self.for_passage = self.for_passage.loc[(self.for_passage['answers'].map(lambda x: len(x)) != 0) & (
+                self.for_passage['passages'].map(lambda x: sum(x['is_selected'])) != 0)].reset_index(drop=True)
+
+
+        self.msmarco = pd.DataFrame(self.for_passage, columns=['query_id', 'question', 'passages', 'answer'])
+        if evaluate_size is not None and len(self.msmarco) > evaluate_size:
+            self.qa_data = self.msmarco[:evaluate_size]
+
+    def ingest(self, retrievals: List[BaseRetrieval], db: BaseDB, ingest_size: Optional[int] = None):
+        """
+        Ingest dataset to retrievals and db.
+        :param retrievals: The retrievals that you want to ingest.
+        :param db: The db that you want to ingest.
+        :param ingest_size: The number of data to ingest. If None, ingest all data.
+        If you want to use context_recall and context_precision metrics, you should ingest all data.
+        """
+
+        make_passages = pd.concat([self.for_passage['query_id'], json_normalize(self.for_passage['passages'].tolist())],
+                                  axis=1)
+        make_passages['passages'] = make_passages.apply(self.__make_passages_and_retrieval_gt, axis=1)
+
+        if ingest_size is not None:
+            passages = [passage for lst_passage in make_passages['passages'][:ingest_size] for passage in lst_passage]
+
+        for retrieval in retrievals:
+            retrieval.ingest(passages)
+        db.create_or_load()
+        db.save(passages)
+
+    def evaluate(self, **kwargs) -> EvaluateResult:
+
+        df = self.qa_data
```

Comment:
you must evaluate passages in the ingested passages. You must validate or guarantee that evaluating passages are in the ingested passages.